app/background/reducers/shieldsPanelReducer.ts

```diff
@@ -0,0 +1,272 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import * as shieldsPanelTypes from '../../constants/shieldsPanelTypes'
+import * as windowTypes from '../../constants/windowTypes'
+import * as tabTypes from '../../constants/tabTypes'
+import * as webNavigationTypes from '../../constants/webNavigationTypes'
+import {
+  setAllowAdBlock,
+  setAllowTrackingProtection,
+  setAllowHTTPSEverywhere,
+  setAllowJavaScript,
+  toggleShieldsValue,
+  requestShieldPanelData
+} from '../api/shieldsAPI'
+import { setBadgeText } from '../api/badgeAPI'
+import { reloadTab } from '../api/tabsAPI'
+import * as shieldsPanelState from '../../state/shieldsPanelState'
+import { State, Tab } from '../../types/state/shieldsPannelState'
+import { Actions } from '../../types/actions/index'
+
+const updateBadgeText = (state: State) => {
+  const tabId: number = shieldsPanelState.getActiveTabId(state)
+  if (state.tabs[tabId]) {
+    const total: string = (state.tabs[tabId].adsBlocked + state.tabs[tabId].trackingProtectionBlocked).toString()
+    setBadgeText(total)
+  }
+}
+
+const focusedWindowChanged = (state: State, windowId: number): State => {
+  if (windowId !== -1) {
+    state = shieldsPanelState.updateFocusedWindow(state, windowId)
+    if (shieldsPanelState.getActiveTabId(state)) {
+      requestShieldPanelData(shieldsPanelState.getActiveTabId(state))
+      updateBadgeText(state)
+    } else {
+      console.warn('no tab id so cannot request shield data from window focus change!')
+    }
+  }
+  return state
+}
+
+const updateActiveTab = (state: State, windowId: number, tabId: number): State => {
+  requestShieldPanelData(tabId)
+  return shieldsPanelState.updateActiveTab(state, windowId, tabId)
+}
+
+export default function shieldsPanelReducer (state: State = { tabs: {}, windows: {}, currentWindowId: -1 }, action: Actions) {
+  switch (action.type) {
+    case webNavigationTypes.ON_BEFORE_NAVIGATION:
+      {
+        if (action.isMainFrame) {
+          state = shieldsPanelState.resetBlockingStats(state, action.tabId)
+        }
+        break
+      }
+    case windowTypes.WINDOW_REMOVED:
+      {
+        state = shieldsPanelState.removeWindowInfo(state, action.windowId)
+        break
+      }
+    case windowTypes.WINDOW_CREATED:
+      {
+        // TODO this is a dirty hack with <Window[]><any> can we improve this?
+        if (action.window.focused || (state.windows as any as Window[]).length === 0) {
```

Comment:
this is actually a bug, it should be Object.keys(state.windows).length === 0

Reply:
fixed